File: app/src/main/java/com/bluzi/babymonitor/log/Log.kt
package com.bluzi.babymonitor.log

/**
 * Pure logging facade — no `android.*` imports, so the protocol layer (xiaomi/, net/, dsp/) can
 * log while staying platform-free (see CLAUDE.md architecture rule). The Android app installs a
 * sink backed by `android.util.Log` at startup; unit tests leave the default no-op sink.
 *
 * One `tag` per subsystem ("login", "cloud", "cs2", "miss", "engine", "ui", "audio", "video").
 * Filter everything on device with: `adb logcat -s BabyMonitor`.
 *
 * Never log secrets — passwords, passToken, serviceToken, ssecurity. Log ids, ips, statuses,
 * and error messages.
 */
object Log {
    enum class Level { DEBUG, INFO, WARN, ERROR }

    fun interface Sink {
        fun log(level: Level, tag: String, message: String, error: Throwable?)
    }

    @Volatile
    private var sink: Sink = Sink { _, _, _, _ -> } // no-op until installed

    fun install(sink: Sink) {
        this.sink = sink
    }

    fun d(tag: String, message: String) = sink.log(Level.DEBUG, tag, message, null)
    fun i(tag: String, message: String) = sink.log(Level.INFO, tag, message, null)
    fun w(tag: String, message: String, error: Throwable? = null) = sink.log(Level.WARN, tag, message, error)
    fun e(tag: String, message: String, error: Throwable? = null) = sink.log(Level.ERROR, tag, message, error)
}

File: windows/src/BabyMonitor.Core/Logging/Log.cs
namespace BabyMonitor.Core.Logging;

/// <summary>
/// Pure logging facade, so the protocol layer can log while staying free of any Windows API.
/// The app installs a sink at start-up; tests leave the default no-op sink.
///
/// One tag per subsystem ("login", "cloud", "cs2", "miss", "engine", "ui", "audio", "video",
/// "update", "app"), exactly as on the phone and the Mac, so a night reads the same on any of them.
///
/// Never log secrets — password, passToken, serviceToken, ssecurity. Log ids, ips, statuses and
/// error messages.
/// </summary>
public static class Log
{
    public enum Level
    {
        Debug,
        Info,
        Warn,
        Error,
    }

    public delegate void Sink(Level level, string tag, string message, Exception? error);

    private static volatile Sink _sink = (_, _, _, _) => { };

    public static void Install(Sink sink) => _sink = sink;

    public static void D(string tag, string message) => _sink(Level.Debug, tag, message, null);

    public static void I(string tag, string message) => _sink(Level.Info, tag, message, null);

    public static void W(string tag, string message, Exception? error = null) =>
        _sink(Level.Warn, tag, message, error);

    public static void E(string tag, string message, Exception? error = null) =>
        _sink(Level.Error, tag, message, error);
}
